appinventor/misc/emulator-support/aiStarter.py

```diff
@@ -71,61 +95,98 @@ def ucheck():
+    shutdown()
+    return {
+        "status": "OK",
+        "version": VERSION
+    }
+
 
 @route('/replstart/:device')
 def replstart(device=None):
-    print "Device = %s" % device
-    subprocess.check_output((PLATDIR + "commands-for-Appinventor/adb -s %s forward tcp:8001 tcp:8001") % device, shell=True, close_fds=True)
-    if re.match('.*emulat.*', device): #  Only fake the menu key for the emulator
-        subprocess.check_output((PLATDIR + "commands-for-Appinventor/adb -s %s shell input keyevent 82") % device, shell=True, close_fds=True)
-    subprocess.check_output((PLATDIR + "commands-for-Appinventor/adb -s %s shell am start -a android.intent.action.VIEW -n edu.mit.appinventor.aicompanion3/.Screen1 --ez rundirect true") % device, shell=True, close_fds=True)
-    response.headers['Access-Control-Allow-Origin'] = '*'
-    response.headers['Access-Control-Allow-Headers'] = 'origin, content-type'
-    return ''
+    print(f"Device = {device}")
+    try:
+        subprocess.check_output(f'"{PLATDIR}adb" -s {device} forward tcp:8001 tcp:8001', shell=True)
+        if re.match('.*emulat.*', device):  # Only fake the menu key for the emulator
+            subprocess.check_output(f'"{PLATDIR}adb" -s {device} shell input keyevent 82', shell=True,)
+        subprocess.check_output(
+            f'"{PLATDIR}adb" -s {device} shell am start -a android.intent.action.VIEW -n edu.mit.appinventor.aicompanion3/.Screen1 --ez rundirect true',
+            shell=True)
+        response.headers['Access-Control-Allow-Origin'] = '*'
+        response.headers['Access-Control-Allow-Headers'] = 'origin, content-type'
+        return ''
+    except subprocess.CalledProcessError as e:
+        print(f"Problem starting companion app : status {e.returncode}\n")
+        return ''
+
 
 def checkrunning(emulator):
-    result = subprocess.check_output(PLATDIR + 'commands-for-Appinventor/adb devices', shell=True, close_fds=True)
-    lines = result.split('\n')
-    for line in lines[1:]:
-        if emulator:
-            m = re.search('^(.*emulator-[1-9]+)\t+device.*', line)
-        else:
-            if re.search('^(.*emulator-[1-9]+)\t+device.*', line): # We are an emulator
-                continue                                           # Skip it
-            m = re.search('^([A-z0-9.:]+.*?)\t+device.*', line)
+    try:
+        result = subprocess.check_output(f'"{PLATDIR}adb" devices', shell=True)
+        lines = result.splitlines()
+        for line in lines[1:]:
+            if line:
+                l = str(line, 'utf-8')  # convert byte to string
+                if emulator:
+                    m = re.search(r'^(emulator-\d+)\s+device.*', l)
+                else:
+                    if re.search(r'^(emulator-\d+)\s+device.*', l):  # We are an emulator
+                        continue  # Skip it
+                    m = re.search(r'^([\w\d]+)\s+device.*', l)
+                if m:
+                    break
         if m:
```

Comment:
I get an UnboundLocalError here when attempting to connect to the emulator.

Reply:
fixed